### ISSUE TITLE
Running inductor codegen from TORCH_LOG="output_code" yields a different result

### ISSUE THREAD

**Issue by xmfan** [pytorch/pytorch#119121]
### 🐛 Describe the bug

I'm investigating an accuracy test failure in https://github.com/pytorch/pytorch/pull/119068. It only happens with compiled autograd + inductor backend, on both cpu and cuda.

I wanted to debug via executing the codegen output from TORCH_LOGS="output_code" with inputs hardcoded in. I was expecting the same results as the compiled autograd run. I'm using a script with hardcoded inputs and weights which I don't think is using any RNG.

repro: https://gist.github.com/xmfan/622c587aeb663aa437a021c626a504be
```python
import torch
from torch._dynamo.utils import maybe_enable_compiled_autograd

def fn():
    model = torch.nn.Sequential(
        torch.nn.Linear(2, 1, bias=False),
        torch.nn.Linear(1, 2, bias=False),
    )
    model[0].weight = torch.nn.Parameter(torch.tensor([[-0.0053,  0.3793]]))
    model[1].weight = torch.nn.Parameter(torch.tensor([[-0.8230],[-0.7359]]))

    x = torch.tensor([[-2.1788,  0.5684], [-1.0845, -1.3986]])

    out = model(x)
    loss = out.sum()
    torch.manual_seed(0)
    loss.backward()

    return (model[0].weight.grad, model[1].weight.grad)

eager_result = fn()
with maybe_enable_compiled_autograd(True):
    compiled_result = fn()

print(eager_result)
# (tensor([[5.0872, 1.2942]]),
#  tensor([[-0.2976], [-0.2976]]))

print(compiled_result)
# (tensor([[5.0872, 1.2942]]),      <-- running inductor codegen directly should output this?
#  tensor([[-1.5589], [-1.5589]]))
```

cpu codegen results differs from compiled_result: https://gist.github.com/xmfan/f66eebca61fea1087ec892910cc251eb
```
# tensor([[5.0872, 1.2942]])
# tensor([[-0.2976], [-0.2976]])
```

cuda codegen results differs from compiled_result: https://gist.github.com/xmfan/e6cabda8e9c01ac7eb3741a026610b1a
```
# tensor([[5.0872, 1.2942]], device='cuda:0')
# tensor([[-0.2976], [-0.2976]], device='cuda:0')
```

### Versions

https://github.com/pytorch/pytorch/pull/119068

cc @ezyang @msaroufim @bdhirsh @anijain2305 @zou3519 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

**Comment by xmfan** [comment]
Found the root cause of this. It's due to both accumulate_grad and inductor trying to reuse tensor storage at the same time. Turning off one or the other prevents this. See annotated cpu codegen:

``` python
# 1. accumulate_grad_ thinks it steals the gradient, but it is only stealing the wrapper object.
# arg3_1's grad points to the same storage as buf1
inductor_ops.accumulate_grad_(arg3_1, reinterpret_tensor(buf1, (2, 1), (1, 1), 0))
# 2. arg3_1 is deleted, but lifetime is kept in aot_module_simplified.forward
del arg3_1
# 3. buf1, who points to the grad of arg3_1, is aliased
buf5 = buf1; del buf1
# 4. buf5 is overwritten i.e. the grad of arg3_1
extern_kernels.mm(buf4, reinterpret_tensor(arg1_1, (2, 1), (1, 1), 0), out=buf5)
```
